File: app/src/main/java/com/aoe/mealsapp/WebActivity.java
package com.aoe.mealsapp;

import android.annotation.SuppressLint;
import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

/**
 * Main activity that contains the WebView that shows the MealsApp website.
 */
public class WebActivity extends AppCompatActivity
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    //
    // CONSTANTS
    //

    private static final String TAG = "## " + WebActivity.class.getSimpleName();

    private static final String PAGE_MAIN = BuildConfig.SERVER_URL;
    private static final String PAGE_LOGIN = BuildConfig.SERVER_URL + "/login";
    private static final String PAGE_TRANSACTIONS = BuildConfig.SERVER_URL + "/accounting/transactions";
    private static final String PAGE_LANGUAGE_SWITCH = BuildConfig.SERVER_URL + "/language-switch";

    private static final String HTTP_USER_AGENT = "MealsApp Android WebView";

    //
    // FIELDS
    //

    private WebView webView;

    //
    // EXTENDS AppCompatActivity
    //

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onCreate() called with: savedInstanceState = [" + savedInstanceState + "]");

        setContentView(R.layout.activity_web);

        setUpAppBar();
        initWebView();

        // set default preference values
        PreferenceManager.setDefaultValues(this, R.xml.preferences, false);

        // register for settings changes
        PreferenceManager.getDefaultSharedPreferences(this)
                .registerOnSharedPreferenceChangeListener(this);

        loadLoginPage();
    }

    /**
     * - open login page with credentials from preferences (send via POST)
     */
    private void loadLoginPage() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        SharedPreferenceKeys keys = SharedPreferenceKeys.getInstance(this);

        String username = sharedPreferences.getString(keys.username, null);
        String password = sharedPreferences.getString(keys.password, null);

        if (username == null) {
            Log.w(TAG, "onCreate: SharedPreferences contains no username. username = \"\"");
            username = "";
        }

        if (password == null) {
            Log.w(TAG, "onCreate: SharedPreferences contains no password. password = \"\"");
            password = "";
        }

        String postData = "_username=" + username + "&_password=" + password;
        webView.postUrl(PAGE_LOGIN, postData.getBytes());
    }

    /**
     * - enable JavaScript
     * - set custom HTTP user agent
     * - set onPageFinished handler that redirects to the LoginActivity if credentials are wrong
     */
    @SuppressLint("SetJavaScriptEnabled")
    private void initWebView() {
        webView = findViewById(R.id.webActivity_webView_webApp);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setUserAgentString(HTTP_USER_AGENT);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                Log.d(TAG, Thread.currentThread().getName() + ": "
                        + "onPageFinished() called with: view = [" + view + "], url = [" + url + "]");

                /* remove trailing slash from URL */

                url = url.replaceAll("/$", "");

                /* unexpected page (e.g. login because bad credentials) -> LoginActivity */

                SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(WebActivity.this);
                SharedPreferenceKeys keys = SharedPreferenceKeys.getInstance(WebActivity.this);

                if (!(url.equals(PAGE_MAIN) || url.equals(PAGE_TRANSACTIONS))) {
                    sharedPreferences.edit().putBoolean(keys.credentialsWereValidated, false).apply();

                    startActivity(new Intent(WebActivity.this, LoginActivity.class));
                    finish();

                } else {
                    sharedPreferences.edit().putBoolean(keys.credentialsWereValidated, true).apply();
                }
            }
        });
    }

    /**
     * - set toolbar as app bar
     * - set AOE logo left to the title
     */
    private void setUpAppBar() {
        Toolbar toolbar = findViewById(R.id.webActivity_toolbar_appBar);
        toolbar.setLogo(R.drawable.logo);
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onCreateOptionsMenu() called with: menu = [" + menu + "]");

        getMenuInflater().inflate(R.menu.menu_main, menu);

        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onDestroy() called");

        /* unregister from settings changes */

        PreferenceManager.getDefaultSharedPreferences(this)
                .unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onOptionsItemSelected() called with: item = [" + item + "]");

        int optionsItemId = item.getItemId();

        switch (optionsItemId) {
            case R.id.mainMenu_alarm:
                AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
                if (alarmManager == null) {
                    Log.e(TAG, "onOptionsItemSelected: Couldn't get AlarmManager. No alarm set.");
                    return true;
                }

                // TODO request code == magic number
                PendingIntent alarmIntent = PendingIntent.getBroadcast(this, 0,
                        new Intent(this, AlarmReceiver.class), 0);
                alarmManager.set(AlarmManager.RTC_WAKEUP,
                        Calendar.getInstance().getTimeInMillis(), alarmIntent);
                break;

            case R.id.mainMenu_account:
                webView.loadUrl(PAGE_TRANSACTIONS);
                return true;

            case R.id.mainMenu_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * If user is not on initial WebView page: Go back in browser history. When user has arrived
     * on initial page: Default behaviour (close activity -> closes app).
     */
    @Override
    public void onBackPressed() {
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onBackPressed() called");

        if (webView.canGoBack())
            webView.goBack();
        else
            super.onBackPressed();
    }

    //
    // IMPLEMENTS SharedPreferences.OnSharedPreferenceChangeListener
    //

    /**
     * Handle all settings changes that are relevant for this activity, i.e. the WebView (e.g.
     * user credentials or website language).
     */
    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Log.d(TAG, Thread.currentThread().getName() + ": "
                + "onSharedPreferenceChanged() called with: sharedPreferences = ["
                + sharedPreferences + "], key = [" + key + "]");

        SharedPreferenceKeys keys = SharedPreferenceKeys.getInstance(this);

        if (key.equals(keys.username) || key.equals(keys.password)) {

            /* load login page with new credentials */

            String username = sharedPreferences.getString(keys.username, null);
            String password = sharedPreferences.getString(keys.password, null);

            // TODO review null handling

            if (username == null) {
                Log.w(TAG, "onSharedPreferenceChanged: username == null. return");
                return;
            }

            if (password == null) {
                Log.w(TAG, "onSharedPreferenceChanged: password == null. return");
                return;
            }

            String postData = "_username=" + username + "&_password=" + password;
            webView.postUrl(PAGE_LOGIN, postData.getBytes());

        } else if (key.equals(keys.language)) {

            /* reload current web page with other language */

            Map<String, String> additionalHttpHeaders = new HashMap<>();
            additionalHttpHeaders.put("Referer", webView.getUrl());
            webView.loadUrl(PAGE_LANGUAGE_SWITCH, additionalHttpHeaders);
        }
    }
}

File: app/src/main/java/com/aoe/mealsapp/WebActivity.kt
package com.aoe.mealsapp

import android.annotation.SuppressLint
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.webkit.*
import com.aoe.mealsapp.settings.Language
import com.aoe.mealsapp.settings.Settings
import com.aoe.mealsapp.settings.SettingsActivity
import kotlinx.android.synthetic.main.activity_web.*
import java.util.*
import java.util.regex.Pattern

class WebActivity : AppCompatActivity() {

    private lateinit var webView: WebView
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout

    // credentials used for last login
    private lateinit var lastUsername: String
    private lateinit var lastPassword: String

    // language returned from last load
    // defaults to GERMAN because the page loads in German by default
    private lateinit var lastLanguage: Language

    // true from onRestoreInstanceState() until first onPageFinished() call
    var restoring = false

    // remember an error in onReceivedError() so that the error page is left
    // untouched in onPageFinished(). Reset it in onPageStarted()
    var errorLoadingPage = false

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onCreate() called with: savedInstanceState = [$savedInstanceState]")

        setContentView(R.layout.activity_web)

        webActivity_toolbar_appBar.setLogo(R.drawable.logo)
        setSupportActionBar(webActivity_toolbar_appBar)

        /* SwipeRefreshLayout & WebView */

        swipeRefreshLayout = webActivity_swipeRefreshLayout_webView
        swipeRefreshLayout.setOnRefreshListener {
            refreshWebView()
        }

        webView = webActivity_webView_webApp
        webView.settings.javaScriptEnabled = true
        webView.settings.userAgentString = HTTP_USER_AGENT
        webView.webViewClient = MyWebViewClient()

        /* */

        val settings = Settings.getInstance(this)

        lastUsername = settings.username
        lastPassword = settings.password
        lastLanguage = settings.language
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onSaveInstanceState() called with: outState = [$outState]")

        webView.saveState(outState)

        outState.putString(STATE_LAST_USERNAME, lastUsername)
        outState.putString(STATE_LAST_PASSWORD, lastPassword)
        outState.putSerializable(STATE_LAST_LANGUAGE, lastLanguage)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onRestoreInstanceState() called with: savedInstanceState = [$savedInstanceState]")

        restoring = true

        webView.restoreState(savedInstanceState)

        lastUsername = savedInstanceState.getString(STATE_LAST_USERNAME)
        lastPassword = savedInstanceState.getString(STATE_LAST_PASSWORD)
        lastLanguage = savedInstanceState.getSerializable(STATE_LAST_LANGUAGE) as Language
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onResume() called")

        /* after return from login / settings screen: check for changes */

        val settings = Settings.getInstance(this)

        val currentUsername = settings.username
        val currentPassword = settings.password
        val currentLanguage = settings.language

        if (lastUsername != currentUsername || lastPassword != currentPassword) {

            /* credentials changed ? save them, re-login */

            lastUsername = currentUsername
            lastPassword = currentPassword

            loadLoginPage(lastUsername!!, lastPassword!!)

        } else if (lastLanguage != currentLanguage) {

            /* language changed ? save it, switch language */

            lastLanguage = currentLanguage

            switchLanguage(webView.url)

        } else if (webView.url !in listOf(PAGE_MAIN, PAGE_TRANSACTIONS)) {

            /* invalid page ? re-login */

            loadLoginPage(lastUsername!!, lastPassword!!)
        }
    }

    override fun onBackPressed() {
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onBackPressed() called")

        if (webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "onOptionsItemSelected() called with: item = [$item]")

        when (item!!.itemId) {
            // debugging: manually trigger alarm
            R.id.mainMenu_alarm ->
                triggerAlarm()

            R.id.mainMenu_account ->
                webView.loadUrl(PAGE_TRANSACTIONS)

            R.id.mainMenu_settings ->
                startActivity(Intent(this, SettingsActivity::class.java))
        }

        return super.onOptionsItemSelected(item)
    }

    // region ### load pages
    //

    private fun loadLoginPage(username: String, password: String) {
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "loadLoginPage() called with: username = [$username], password = [$password]")

        webView.clearHistory()

        val postData = ("_username=$username&_password=$password")
        webView.postUrl(PAGE_LOGIN, postData.toByteArray())
    }

    private fun switchLanguage(targetUrl: String) {
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "switchLanguage() called with: targetUrl = [$targetUrl]")

        webView.loadUrl(PAGE_LANGUAGE_SWITCH, mutableMapOf("Referer" to targetUrl))
    }

    private fun refreshWebView() {
        Log.d(TAG, Thread.currentThread().name + " ### " +
                "refreshWebView() called")

        webView.reload()
    }

    //
    // endregion

    // region ### debug
    //

    private fun triggerAlarm() {
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager

        val alarmIntent = PendingIntent.getBroadcast(this, 0,
                Intent(this, AlarmReceiver::class.java), 0)
        alarmManager.set(AlarmManager.RTC_WAKEUP,
                Calendar.getInstance().timeInMillis, alarmIntent)
    }

    //
    // endregion

    inner class MyWebViewClient : WebViewClient() {

        override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
            super.onPageStarted(view, url, favicon)
            Log.d(TAG, Thread.currentThread().name + " ### " +
                    "onPageStarted() called with: view = [$view], url = [$url], favicon = [$favicon]")

            swipeRefreshLayout.isRefreshing = true

            errorLoadingPage = false
        }

        override fun onPageFinished(view: WebView, url: String) {
            super.onPageFinished(view, url)
            Log.d(TAG, Thread.currentThread().name + " ### " +
                    "onPageFinished() called with: view = [$view], url = [$url]")

            if (errorLoadingPage)
                return

            swipeRefreshLayout.isRefreshing = false

            if (restoring) {
                restoring = false
            } else {

                /* read language from cookies & save it */

                // TODO check
                var cookieString = CookieManager.getInstance().getCookie(PAGE_MAIN)
                if (cookieString != null) {
                    cookieString = cookieString.replace(" ", "")
                    val cookies = Pattern.compile(";").split(cookieString)
                    for (cookie in cookies) {
                        val cookieParts = cookie.split("=")
                        if (cookieParts[0] == "locale") {
                            lastLanguage = when (cookieParts[1]) {
                                "de" -> Language.GERMAN
                                "en" -> Language.ENGLISH
                                else -> Language.ENGLISH
                            }
                        }
                    }
                } else {
                    lastLanguage = Language.ENGLISH
                }

                /* wrong page (invalid credentials) ? show LoginActivity */

                if (webView.url !in listOf(PAGE_MAIN, PAGE_TRANSACTIONS)) {
                    startActivity(Intent(this@WebActivity, LoginActivity::class.java))
                } else {

                    /* valid page but wrong langauge ? switch language */

                    val currentLanguage = Settings.getInstance(this@WebActivity).language

                    if (lastLanguage != currentLanguage) {

                        /* language changed ? save it, switch language */

                        lastLanguage = currentLanguage

                        switchLanguage(webView.url)
                    }
                }
            }
        }


        override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
            super.onReceivedError(view, request, error)
            Log.d(TAG, Thread.currentThread().name + " ### " +
                    "onReceivedError() called with: view = [$view], request = [$request], error = [$error]")

            swipeRefreshLayout.isRefreshing = false

            if (restoring) {
                restoring = false
            }

            errorLoadingPage = true
        }
    }

    private companion object {

        private const val TAG = "WebActivity"

        private const val PAGE_MAIN = BuildConfig.SERVER_URL + "/"
        private const val PAGE_LOGIN = BuildConfig.SERVER_URL + "/login"
        private const val PAGE_TRANSACTIONS = BuildConfig.SERVER_URL + "/accounting/transactions"
        private const val PAGE_LANGUAGE_SWITCH = BuildConfig.SERVER_URL + "/language-switch"

        private const val HTTP_USER_AGENT = "MealsApp Android WebView"

        private const val STATE_LAST_USERNAME = "STATE_LAST_USERNAME"
        private const val STATE_LAST_PASSWORD = "STATE_LAST_PASSWORD"
        private const val STATE_LAST_LANGUAGE = "STATE_LAST_LANGUAGE"
    }
}
